lib/core/src/sdk.rs

```diff
@@ -981,8 +1017,77 @@ impl LiquidSdk {
             }
 
             // If no MRH found, perform usual swap
-            None => self.send_payment_via_swap(invoice, fees_sat).await,
+            None => {
+                self.send_payment_via_swap(
+                    invoice,
+                    &bolt11_invoice.payment_hash().to_string(),
+                    description,
+                    amount_sat,
+                    fees_sat,
+                )
+                .await
+            }
+        }
+    }
+
+    async fn pay_bolt12_invoice(
+        &self,
+        offer: &LNOffer,
+        user_specified_receiver_amount_sat: u64,
+        invoice: &str,
+        fees_sat: u64,
+    ) -> Result<SendPaymentResponse, PaymentError> {
+        let invoice_parsed = utils::parse_bolt12_invoice(invoice)?;
+        let invoice_signing_pubkey = invoice_parsed.signing_pubkey().to_hex();
+
+        // Check if the invoice is signed by same key as the offer
+        match &offer.signing_pubkey {
+            None => {
+                ensure_sdk!(
+                    &offer
+                        .paths
+                        .iter()
+                        .filter_map(|path| path.blinded_hops.last())
+                        .any(|last_hop| &invoice_signing_pubkey == last_hop),
+                    PaymentError::invalid_invoice(
+                        "Invalid Bolt12 invoice signing key when using blinded path"
+                    )
+                );
+            }
+            Some(offer_signing_pubkey) => {
+                ensure_sdk!(
+                    offer_signing_pubkey == &invoice_signing_pubkey,
+                    PaymentError::invalid_invoice("Invalid Bolt12 invoice signing key")
+                );
+            }
         }
+
+        let receiver_amount_sat = invoice_parsed.amount_msats() / 1_000;
+        ensure_sdk!(
+            receiver_amount_sat == user_specified_receiver_amount_sat,
+            PaymentError::invalid_invoice("Invalid Bolt12 invoice amount")
+        );
+        if let Some(Amount::Bitcoin { amount_msat }) = &offer.min_amount {
+            ensure_sdk!(
+                receiver_amount_sat >= amount_msat / 1_000,
+                PaymentError::invalid_invoice("Invalid Bolt12 invoice amount: below offer minimum")
+            );
+        }
```

Comment:
Perhaps we should be doing this in the prepare step, as the invoice creation will likely fail if the request amount in below the offer min amount

Reply:
Good point, done: 9920648b1c6a1cd108a4edc8373eeff1f1aec539